File: client/grao-a-grao/src/types/stock_in.ts
import { InvalidBuyPriceError, InvalidPackagingQuantityError, InvalidTotalQuantityError, MissingItemIdError, MissingPackagingIdError, MissingPackagingsError, MissingStockInIdError, NoItemsError, NonFractionablePackagingError } from "@/errors/stockInValidation";
import { ItemModel, ItemResponse, normalizeItemResponse } from "./item";
import { ItemPackagingModel, ItemPackagingResponse, normalizeItemPackagingResponse } from "./item_packaging";

// Frontend model
export interface StockInModel {
  id?: number;
  status?: string;
  items: StockInItemModel[];
  created_at?: string;
  updated_at?: string;
  finalized_at?: string;
}

export interface StockInItemModel {
  id?: number;
  item: ItemModel;
  buy_price: number;
  total_quantity: number;
  packagings: StockInPackagingModel[];
}

export interface StockInPackagingModel {
  id?: number;
  item_packaging: ItemPackagingModel;
  quantity: number;
}

// Request DTOs
export interface CreateStockInRequest {
  items: CreateStockInItemRequest[];
}

export interface CreateStockInItemRequest {
  item_id: number;
  buy_price: number;
  total_quantity: number;
  packagings: CreateStockInPackagingRequest[];
}

export interface CreateStockInPackagingRequest {
  item_packaging_id: number;
  quantity: number;
}

export interface UpdateStockInRequest {
  id: number;
  items: UpdateStockInItemRequest[];
}

export interface UpdateStockInItemRequest {
  id?: number;
  item_id: number;
  buy_price: number;
  total_quantity: number;
  packagings: UpdateStockInPackagingRequest[];
}

export interface UpdateStockInPackagingRequest {
  id?: number;
  item_packaging_id: number;
  quantity: number;
}

// Response DTOs
export interface StockInResponse {
  id: number;
  status: string;
  items: StockInItemResponse[];
  created_at: string;
  updated_at: string;
  finalized_at: string;
}

export interface StockInItemResponse {
  id: number;
  item: ItemResponse;
  buy_price: number;
  total_quantity: number;
  packagings: StockInPackagingResponse[];
}

export interface StockInPackagingResponse {
  id: number;
  item_packaging: ItemPackagingResponse;
  quantity: number;
}

/**
 * Safely normalize API response into frontend model,
 * handling potential null or undefined arrays.
 */
export function normalizeStockInResponse(
  res: StockInResponse
): StockInModel {
  return {
    id: res.id,
    status: res.status,
    created_at: res.created_at ?? undefined,
    updated_at: res.updated_at ?? undefined,
    finalized_at: res.finalized_at ?? undefined,
    items: (res.items ?? []).map((item: StockInItemResponse): StockInItemModel => ({
      id: item.id,
      item: normalizeItemResponse(item.item),
      buy_price: item.buy_price,
      total_quantity: item.total_quantity,
      packagings: (item.packagings ?? []).map(pkg => ({
        id: pkg.id,
        item_packaging: normalizeItemPackagingResponse(pkg.item_packaging),
        quantity: pkg.quantity,
      })),
    })),
  };
}


// Convert frontend model to Create request
export function toCreateStockInRequest(model: StockInModel): CreateStockInRequest {
  if (!model.items || model.items.length === 0) throw new NoItemsError();

  return {
    items: model.items.map(item => {
      if (item.item.id === undefined) throw new MissingItemIdError();

      if (item.buy_price <= 0) throw new InvalidBuyPriceError();

      if (item.total_quantity <= 0) throw new InvalidTotalQuantityError();

      // Validation conditioned to item fractionability
      if (!item.item.is_fractionable) {
        // non fractionable
        if (item.packagings.length > 0) throw new NonFractionablePackagingError();
      } else {
        // fractionable
        if (!item.packagings || item.packagings.length === 0) throw new MissingPackagingsError();
      }


      return {
        item_id: item.item.id,
        buy_price: item.buy_price,
        total_quantity: item.total_quantity,
        packagings: item.packagings.map(pkg => {

          if (pkg.item_packaging.id === undefined) throw new MissingPackagingIdError();
          if (pkg.quantity <= 0) throw new InvalidPackagingQuantityError();
          
          return {
            item_packaging_id: pkg.item_packaging.id,
            quantity: pkg.quantity,
          };
        }),
      };
    }),
  };
}

// Convert frontend model to Update request
export function toUpdateStockInRequest(model: StockInModel): UpdateStockInRequest {
  if (model.id === undefined) throw new MissingStockInIdError();

  if (!model.items || model.items.length === 0) throw new NoItemsError();

  return {
    id: model.id,
    items: model.items.map(item => {

      if (item.item.id === undefined) throw new MissingItemIdError();

      if (item.buy_price <= 0) throw new InvalidBuyPriceError();

      if (item.total_quantity <= 0) throw new InvalidTotalQuantityError();

      // Validation conditioned to item fractionability
      if (!item.item.is_fractionable) {
        // non fractionable
        if (item.packagings.length > 0) throw new NonFractionablePackagingError();
      } else {
        // fractionable
        if (!item.packagings || item.packagings.length === 0) throw new MissingPackagingsError();
      }

      return {
        id: item.id,
        item_id: item.item.id,
        buy_price: item.buy_price,
        total_quantity: item.total_quantity,
        packagings: item.packagings.map(pkg => {

          if (pkg.item_packaging.id === undefined) throw new MissingPackagingIdError();
          if (pkg.quantity <= 0) throw new InvalidPackagingQuantityError();

          return {
            id: pkg.id,
            item_packaging_id: pkg.item_packaging.id,
            quantity: pkg.quantity,
          };
        }),
      };
    }),
  };
}

File: pkg/dto/mapper/stock_in.go
package mapper

import (
	"github.com/IlfGauhnith/GraoAGrao/pkg/dto/request"
	"github.com/IlfGauhnith/GraoAGrao/pkg/dto/response"
	"github.com/IlfGauhnith/GraoAGrao/pkg/model"
)

func CreateStockInToModel(r *request.CreateStockInRequest) *model.StockIn {
	var items []model.StockInItem

	for _, itr := range r.Items {
		var packagings []model.StockInPackaging
		for _, p := range itr.Packagings {
			packagings = append(packagings, model.StockInPackaging{
				ItemPackaging: model.ItemPackaging{ID: p.ItemPackagingID},
				Quantity:      p.Quantity,
			})
		}

		stockInItem := model.StockInItem{
			Item:          model.Item{ID: itr.ItemID},
			BuyPrice:      itr.BuyPrice,
			TotalQuantity: itr.TotalQuantity,
			Packagings:    packagings,
		}

		items = append(items, stockInItem)
	}

	return &model.StockIn{
		Items: items,
	}
}

func UpdateStockInToModel(r *request.UpdateStockInRequest) *model.StockIn {
	var items []model.StockInItem

	for _, itr := range r.Items {
		var packagings []model.StockInPackaging
		for _, p := range itr.Packagings {
			packagings = append(packagings, model.StockInPackaging{
				ID:            getID(p.ID),
				ItemPackaging: model.ItemPackaging{ID: p.ItemPackagingID},
				Quantity:      p.Quantity,
			})
		}

		stockInItem := model.StockInItem{
			ID:            getID(itr.ID),
			StockInID:     r.ID,
			Item:          model.Item{ID: itr.ItemID},
			BuyPrice:      itr.BuyPrice,
			TotalQuantity: itr.TotalQuantity,
			Packagings:    packagings,
		}

		items = append(items, stockInItem)
	}

	return &model.StockIn{
		ID:    r.ID,
		Items: items,
	}
}

func ToStockInResponse(m *model.StockIn) *response.StockInResponse {
	var items []response.StockInItemResponse

	for _, i := range m.Items {
		var packagings []response.StockInPackagingResponse
		for _, p := range i.Packagings {
			packagings = append(packagings, response.StockInPackagingResponse{
				ID:            p.ID,
				ItemPackaging: ToItemPackagingResponse(&p.ItemPackaging),
				Quantity:      p.Quantity,
			})
		}

		items = append(items, response.StockInItemResponse{
			ID:            i.ID,
			Item:          ToItemResponse(&i.Item),
			BuyPrice:      i.BuyPrice,
			TotalQuantity: i.TotalQuantity,
			Packagings:    packagings,
		})
	}

	return &response.StockInResponse{
		ID:        m.ID,
		Status:    m.Status,
		Items:     items,
		CreatedAt: m.CreatedAt,
		UpdatedAt: m.UpdatedAt,
	}
}

func getID(id *uint) uint {
	if id != nil {
		return *id
	}
	return 0
}
